### PR TITLE
uint64 to number64

### PR DESCRIPTION
**Pull request by wemeetagain**
Change Epoch, Shard, Slot, ValidatorIndex, genesisTime, depositIndex
to number64.

This relies on the chain module PR and the new ssz library. It won't pass tests until we upgrade ssz b/c our current version of ssz doesn't support number64 types.

### CODE REVIEW

**Comment by GregTheGreek** [comment]
This is an awesome pr i love it!
Is there even a need to call it `number64`?

**Comment by wemeetagain** [comment]
there are a few things named number64
- the typescript type - this doesn't need to be named `number64`, I named it this to mirror the `uint64` type, but we can use something else
- the ssz type variable name - this doesn't need to be named `number64`, but it does need to be named the same as the typescsript type (our type consistency tests make sure of this)
- the ssz type variable value - this either needs to be either `"number64"` or `{type: Type.uint, bytelength: 8, offset: 0, useNumber: true}` so that ssz knows how to de/serialize it

**Comment by GregTheGreek** [comment]
ahhhh i gothcya, no i totally agree it does make sense! Theres some tests failing due to the changes not sure why. i can look into those if you want?

**Comment by wemeetagain** [comment]
Tests are failing due to still using the old ssz, which doesn't support `number64`. We'll need to upgrade to the typescript rewrite to get `number64` support.

**Comment by GregTheGreek** [comment]
Do we want to wait for SSZ?

**Comment by wemeetagain** [comment]
Yeah, we should probably wait for a new ssz, so we can keep our tests passing.

**Comment by GregTheGreek** [comment]
@wemeetagain Unfortunately, we need to wait for npm to get back to us -- ill add you to the exisiting team and we can go from there.

**Comment by GregTheGreek** [comment]
@wemeetagain sooooo we got locked out of npm completely... I'm trying to get on call with them...

**Comment by GregTheGreek** [comment]
With an update to SSZ and a conflict fix we should be good :)

**Comment by wemeetagain** [comment]
This should be good to go~

File: src/chain/helpers/stateTransitionHelpers.ts [hunk 1]
```diff
@@ -380,11 +382,11 @@ export function getCrosslinkCommitteesAtSlot(state: BeaconState, slot: Slot, reg
...
     shufflingEpoch = nextEpoch;
 
-    const epochsSinceLastRegistryUpdate: BN = currentEpoch.sub(state.validatorRegistryUpdateEpoch);
+    const epochsSinceLastRegistryUpdate = currentEpoch - state.validatorRegistryUpdateEpoch;
     if (registryChange) {
       seed = generateSeed(state, nextEpoch);
-      shufflingStartShard = (state.currentShufflingStartShard.addn(currentCommitteesPerEpoch)).umod(new BN(SHARD_COUNT));
-    } else if (epochsSinceLastRegistryUpdate.gtn(1) && isPowerOfTwo(epochsSinceLastRegistryUpdate)) {
+      shufflingStartShard = (state.currentShufflingStartShard + currentCommitteesPerEpoch) % SHARD_COUNT;
+    } else if (epochsSinceLastRegistryUpdate < 1 && isPowerOfTwo(epochsSinceLastRegistryUpdate)) {
```
Comment (GregTheGreek) [inline src/chain/helpers/stateTransitionHelpers.ts]: should be `>`

File: src/chain/index.ts [hunk 2]
```diff
@@ -109,8 +120,8 @@ export class BeaconChain extends EventEmitter {
     // TODO: implement
     
     // The node's Unix time is greater than or equal to state.genesis_time + (block.slot - GENESIS_SLOT) * SECONDS_PER_SLOT.
-    const stateSlotTime = state.genesisTime.add(block.slot.sub(GENESIS_SLOT).muln(SECONDS_PER_SLOT));
-    if (!(new BN(Date.now())).gte(stateSlotTime)) {
+    const stateSlotTime = state.genesisTime + (block.slot - GENESIS_SLOT) * SECONDS_PER_SLOT;
```
Comment (GregTheGreek) [inline src/chain/index.ts]: `const stateSlotTime = state.genesisTime + ((block.slot - GENESIS_SLOT) * SECONDS_PER_SLOT);`
I think you made a mistake with brackets

File: src/constants/constants.ts [hunk 3]
```diff
@@ -38,6 +27,18 @@ export const ACTIVATION_EXIT_DELAY = 2 ** 2; // epochs || 25.6 minutes
 export const EPOCHS_PER_ETH1_VOTING_PERIOD = 2 ** 4; // epochs || ~1.7 hours
 export const MIN_VALIDATOR_WITHDRAWAL_DELAY = 2 ** 8; // epochs || ~27 hours
 
+// Initial values
+export const GENESIS_FORK_VERSION	= new BN(0);
```
Comment (GregTheGreek) [inline src/constants/constants.ts]: does this actually need to be BN?

File: src/constants/constants.ts [hunk 4]
```diff
@@ -38,6 +27,18 @@ export const ACTIVATION_EXIT_DELAY = 2 ** 2; // epochs || 25.6 minutes
...
+export const GENESIS_SLOT = 0;
+export const SLOT_OFFSET = 2 ** 32; // 2 ** 32
```
Comment (GregTheGreek) [inline src/constants/constants.ts:33]: 2 ** 63?

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
Few questions below otherwise LGTM

File: src/constants/constants.ts (hunk 3 above)
Comment (wemeetagain) [reply src/constants/constants.ts]: Good question. I wasn't sure about fork versions, so I just left them as BN.
If we can confirm that we're starting at 0 and incrementing from there, then we could move to `number`.

File: src/constants/constants.ts (hunk 4 above)
Comment (wemeetagain) [reply src/constants/constants.ts:33]: Ah yeah, I was already looking ahead, lol

File: src/chain/index.ts (hunk 2 above)
Comment (wemeetagain) [reply src/chain/index.ts]: should be the same order of operations ;) I can add the parens for more readability though.

File: src/constants/constants.ts (hunk 4 above)
Comment (wemeetagain) [reply src/constants/constants.ts:33]: actually, there's a problem using 2 ** 63 right now. Our ssz currently assumes a `number` offset.
We should change it in any case to allow for a BN offset, but for now this variable has to be a number.

File: src/constants/constants.ts (hunk 3 above)
Comment (GregTheGreek) [reply src/constants/constants.ts]: Just asked, it's always 4 bytes so we should be good

Comment (GregTheGreek) [reply src/constants/constants.ts]: I.e number should work

File: src/constants/constants.ts (hunk 4 above)
Comment (GregTheGreek) [reply src/constants/constants.ts:33]: Ahhhh...

File: src/constants/constants.ts (hunk 3 above)
Comment (GregTheGreek) [reply src/constants/constants.ts]: And they won't necessarily be incremental, rather 4 bytes and we'll need to do some bitwise shifting, the rational was due to potential forks in the future... Although we can always patch this down the road.

It makes more sense as number imo but ultimately doesn't matter

Comment (wemeetagain) [reply src/constants/constants.ts]: Ok! hooray! will update

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**